### PR TITLE
Adds new resource cloudflare_byo_ip_prefix

### PR DESCRIPTION
**Pull request by cehrig**
This PR introduces a new resource `cloudflare_ip_prefix` for managing
- IP Prefix descriptions via the `description` attribute
- IP Prefix Dynamic Advertisement Status via the `advertisement` attribute

The resource needs `account_id` to be set in the provider configuration.

It uses `GetPrefix` and `GetAdvertisementStatus` for reading IP prefix information + `UpdatePrefixDescription` and `UpdateAdvertisementStatus` for writing updates from `cloudflare-go`.

@jacobbednarz I'd appreciate if you could take a look if that's going into the right direction before I proceed. Some remarks:
- I'm not using a boolean for the advertisement status, as there's no reliable alternative for `GetOK` to check if an optional boolean has been set and `GetOKExists` is marked as deprecated.
- Writing a test for updating the advertisement status is going to be complicated since there's a conservative rate-limit on the API

**TODO**
- Update cloudflare-go to the latest version
- Write acceptance test for updating the prefix description
- Write doc

### CODE REVIEW

**Comment by cehrig** [comment]
> > I'm not using a boolean for the advertisement status, as there's no reliable alternative for GetOK to check if an optional boolean has been set and GetOKExists is marked as deprecated.
> 
> Have you seen [`cloudflare_record` proxied status](https://github.com/terraform-providers/terraform-provider-cloudflare/blob/6e3af4413715aada7cc00d12fbe7be6aabbff592/cloudflare/resource_cloudflare_record.go#L247-L251)? Looks like a similar use case to what you're after here that works.
> 

Looking at `cloudflare_record`, that seems to work fine when a default value is set in the schema. I'm afraid we can't set a default value for the advertisement status. The use-case here is if someone creates the resource locally and just wants to "terraform" the description, not specifying the `advertisement` attribute.

> > Writing a test for updating the advertisement status is going to be complicated since there's a conservative rate-limit on the API
> 
> Let's give it a go and see what we need to do. I don't really want functionality without tests to make it into the provider as we've been there and it makes maintainer life especially hard. If we hit rate limits, we can chat to Cloudflare about some next steps.

Fully agree. Acceptance-Testing the prefix description is completely fine. The only concern here is testing the advertisement status. I'll check internally if there might be a way.

File: cloudflare/resource_cloudflare_ip_prefix.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+				ValidateFunc: validation.StringInSlice([]string{"on", "off"}, false),
+				Computed:     true,
+				Optional:     true,
+			},
+		},
+	}
+}
+
+func resourceCloudflareIPPrefixCreate(d *schema.ResourceData, meta interface{}) error {
+	if err := resourceCloudflareIPPrefixRead(d, meta); err != nil {
```
[the remark this answers is not in the document]
Comment (cehrig) [reply cloudflare/resource_cloudflare_ip_prefix.go]: Good point. Updating now first, then syncing with `Read`

File: cloudflare/resource_cloudflare_ip_prefix.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	d.Set("advertisement", stringFromBool(advertisementStatus.Advertised))
+
+	return nil
+}
+
+func resourceCloudflareIPPrefixUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	prefixID := d.Get("prefix_id").(string)
+	d.SetId(prefixID)
```
[the remark this answers is not in the document]
Comment (cehrig) [reply cloudflare/resource_cloudflare_ip_prefix.go]: Thx! Updated.

[the remark this answers is not in the document]
Comment (jacobbednarz) [reply cloudflare/resource_cloudflare_ip_prefix.go]: Most likely, yes.

File: cloudflare/provider.go [hunk 3]
```diff
@@ -114,6 +114,7 @@ func Provider() terraform.ResourceProvider {
 			"cloudflare_custom_ssl":                  resourceCloudflareCustomSsl(),
 			"cloudflare_filter":                      resourceCloudflareFilter(),
 			"cloudflare_firewall_rule":               resourceCloudflareFirewallRule(),
+			"cloudflare_ip_prefix":                   resourceCloudflareIPPrefix(),
```
[the remark this answers is not in the document]
Comment (tarnfeld) [reply cloudflare/provider.go:117]: That's fair. This particular feature set is also knows as BYOIP (Bring Your Own IP) and is an account level feature, that sits at a level across all services. If you want to be more specific about the prefixes and other things in /addressing then perhaps `cloudflare_byo_ip_prefix` or `cloudflare_addr_ip_prefix` (addressing) might suit.

**Comment by tarnfeld** [comment]
@cehrig unfortunately there isn't a way we can relax that rate limit without essentially performing a noop if the call is within a certain time frame (15 minutes, for example). Route changes take time to converge and are disruptive to the internet, so we have to be quite conservative about how frequently we allow external parties to change bgp status.

One recommendation would be to check the latest advertised modified timestamp, and only perform the PATCH request if it is beyond the 15 minute time window. I'm sorry I can't make this any easier for your testing. If it makes you feel better, we have the same issue internally. It's important to test frequently, but we can't disrupt the internet for the sake of our tests... and if you're testing the API but not actually changing BGP -- then it's not really a good test.

Can you mark the test case as Skipped somehow when you reach this condition? Perhaps you can at least highlight this event with `t.Skip(msg)`

File: cloudflare/provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobbednarz) [reply cloudflare/provider.go:117]: I think `cloudflare_byo_ip_prefix` is probably a great middle ground. It's specific, conveys what it intended use is for and doesn't pigeon hole us if new features land on the horizon and mention anything about IP prefixes 😛

**Comment by jacobbednarz** [comment]
Your testing comments make total sense, thanks @tarnfeld. Given the difficulty in testing this, I think we may just need to rely on manual use for the Terraform side of things and unit tests within `cloudflare-go` to ensure we're doing everything at that level correct. It's a shame there isn't a staging internet we could try these things on 😂

File: cloudflare/provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cehrig) [reply cloudflare/provider.go]: Awesome. Updated.

File: cloudflare/resource_cloudflare_ip_prefix.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cehrig) [reply cloudflare/resource_cloudflare_ip_prefix.go]: Done. Updated.

**Comment by cehrig** [comment]
I have at least added a test for the prefix description. It requires `CLOUDFLARE_BYO_IP_PREFIX_ID` set to the prefix ID. If that looks good, I'd setup a doc for the website.

```
cehrig@che-box ~/go/src/github.com/terraform-providers/terraform-provider-cloudflare $ go test -count=1 -v -timeout 120s github.com/terraform-providers/terraform-provider-cloudflare/cloudflare -run TestAccCloudflareBYOIPPrefix
=== RUN   TestAccCloudflareBYOIPPrefix
=== PAUSE TestAccCloudflareBYOIPPrefix
=== CONT  TestAccCloudflareBYOIPPrefix
--- PASS: TestAccCloudflareBYOIPPrefix (10.86s)
PASS
ok      github.com/terraform-providers/terraform-provider-cloudflare/cloudflare 10.874s
```

File: cloudflare/provider_test.go [hunk 4]
```diff
@@ -74,6 +74,24 @@ func testAccPreCheckLogpushToken(t *testing.T) {
...
+	}
+
+	if v := os.Getenv("CLOUDFLARE_API_KEY"); v == "" {
+		t.Fatal("CLOUDFLARE_API_KEY must be set for acceptance tests")
+	}
+
+	if v := os.Getenv("CLOUDFLARE_ACCOUNT_ID"); v == "" {
+		t.Fatal("CLOUDFLARE_ACCOUNT_ID must be set for this acceptance test")
+	}
+
```
Comment (jacobbednarz) [inline cloudflare/provider_test.go:89]: You don't need to include these in here as we already have methods for them which can be referenced together.

```suggestion 
```

File: cloudflare/resource_cloudflare_byo_ip_prefix_test.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+func TestAccCloudflareBYOIPPrefix(t *testing.T) {
+	t.Parallel()
+	prefixID := os.Getenv("CLOUDFLARE_BYO_IP_PREFIX_ID")
+
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_byo_ip_prefix.%s", rnd)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheckBYOIPPrefix(t) },
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_byo_ip_prefix_test.go:19]: `testAccPreCheck` takes care of ensuring the credentials are in place
`testAccPreCheckAccount` ensures you have the account details setup

```suggestion
		PreCheck:  func() {
			testAccPreCheck(t)
			testAccPreCheckAccount(t)
			testAccPreCheckBYOIPPrefix(t) 
		},
```

File: cloudflare/resource_cloudflare_byo_ip_prefix_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		},
+	})
+}
+
+func testAccCheckCloudflareBYOIPPrefixConfig(prefixID, description, name string) string {
+	return fmt.Sprintf(`
+				resource "cloudflare_byo_ip_prefix" "%[3]s" {
+					prefix_id = "%[1]s"    
+					description = "%[2]s"
+				}`, prefixID, description, name)
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_byo_ip_prefix_test.go:45]: Can we format this JSON to not have really big indents?

File: website/docs/r/byo_ip_prefix.html.markdown [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `prefix_id` - (Required) The assigned Bring-Your-Own-IP prefix ID.
+* `description` - (Optional) The description of the prefix.
+* `advertisement` - (Optional) Whether or not the prefix shall be announced. A prefix can be activated or deactivated once every 15 minutes. Valid values: `on` or `off`.
```
Comment (jacobbednarz) [inline website/docs/r/byo_ip_prefix.html.markdown:29]:
```suggestion
* `advertisement` - (Optional) Whether or not the prefix shall be announced. A prefix can be activated or deactivated once every 15 minutes (attempting more regular updates will trigger rate limiting). Valid values: `on` or `off`.
```

File: website/docs/r/byo_ip_prefix.html.markdown [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
```
Comment (jacobbednarz) [inline website/docs/r/byo_ip_prefix.html.markdown:30]:
```suggestion
```

File: cloudflare/resource_cloudflare_byo_ip_prefix_test.go (hunk 6 above)
Comment (cehrig) [reply cloudflare/resource_cloudflare_byo_ip_prefix_test.go]: I removed the indentation, I hope that meant with formatting? :-)

**Comment by jacobbednarz** [comment]
Thanks a bunch for persisting to get this one over the line @cehrig! You've done a great job here ⭐

**Comment by cehrig** [comment]
> Thanks a bunch for persisting to get this one over the line @cehrig! You've done a great job here

@jacobbednarz I have to Thank You. I'll keep in mind to run the full test suite in the future ... You rock, Thanks a lot! :-)

**Pull request merged**